### PR TITLE
[#10938] Support copying courses 

### PR DESCRIPTION
**Pull request by LeeBadal**
Part of  #10938 

**Outline of Solution**
* Implements a copy course option for course instructors in the same fashion as copy sessions
* Includes an option to chose specific sessions
* Includes basic front end tests.
NOTE: Is missing E2E tests, did not manage to implement tests that passed reliably. Most likely need assistance if this would be requested.

Screenshots of use case below:
Screenshot from 2021-03-05 18-55-11
Screenshot from 2021-03-05 19-00-07
Screenshot from 2021-03-05 19-01-12

### CODE REVIEW

**Comment by damithc** [comment]
Some suggestions for the modal dialog:
* Title need to contain 'and its sessions'. It's implied.
* Also show which course (id, name) is being copied
* Also allow specifying name and timezone ? Possibly reuse the code from 'create new course' feature?
* If you use a table for the session list, you can use a tick box in the top row for 'select/unselect all' behavior, which is more common UX pattern these days
* Avoid the word 'Choose' as it hints of a choice between options

**Comment by Derek-Hardy** [comment]
@LeeBadal Lint tests are failing, which suggests some of your coding style do not conform with the project standard. Try to run `./gradlew lint` or `npm run lint` to inspect and fix them.

**Comment by Olle251** [comment]
The coding style is fixed.

The modal for copying a course is now updated to look like this:

Screenshot 2021-03-08 at 13 18 39

We did this as an assignment in school and we hope our solution is useful for this project. 

We will probably not have time to fix more on this issue, so if there is more to be fixed someone else is welcome to take over from here.

**Comment by damithc** [comment]
Good progress!
Here's an example of how to use a 'select all' checkbox
image

It doesn't need a label if you place is appropriately.

If not using a table, perhaps you can try something like this?

> [  ] **Sessions to copy:**
> [  ] session 1
> [  ] session 2

**Comment by Olle251** [comment]
Changed it to look like this now:

Screenshot 2021-03-09 at 19 31 26

**Comment by damithc** [comment]
> Changed it to look like this now:
> 
> Screenshot 2021-03-09 at 19 31 26

Nice. UI looks OK to me now, but that's just me. Feedback from others welcome.

**Comment by damithc** [comment]
> UI looks OK to me now, but that's just me. Feedback from others welcome.

Actually, you can consider something like this:

> ### Copy course
> ----
> **Copy from**: [SE101-2021] Software Engineering
>
> **New course**:
> Course ID: [   ]
> Course name: [  ]
> Timezone: [  ]
> ...

**Comment by Olle251** [comment]
OK, now it looks like this:
Screenshot 2021-03-10 at 12 13 49

**Comment by damithc** [comment]
> OK, now it looks like this:
> Screenshot 2021-03-10 at 12 13 49

👍

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import { TimezoneService } from '../../../services/timezone.service';
+import { FeedbackSession } from '../../../types/api-output';
+import { COURSE_ID_MAX_LENGTH } from '../../../types/field-validator';
+
+interface Timezone {
+  id: string;
+  offset: string;
+}
+
+const formatTwoDigits: Function = (n: number): string => {
```
Comment (daongochieu2810) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:13]: With ES2017, we can do 
`const zeroPad = (num, places) => String(num).padStart(2, '0')`
[source](https://stackoverflow.com/questions/2998784/how-to-output-numbers-with-leading-zeros-in-javascript)

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts [hunk 2]
```diff
@@ -246,6 +257,132 @@ export class InstructorCoursesPageComponent implements OnInit {
...
+   * Creates a copy of a course including all sessions.
+   */
+  onCopy(courseId: string, courseName: string, timeZone: string): void {
+    if (!courseId) {
+      this.statusMessageService.showErrorToast(`Course ${courseId} is not found!`);
+      return;
+    }
+    this.feedbackSessionsService.getFeedbackSessionsForInstructor(courseId).subscribe((response: FeedbackSessions) => {
+      const modalRef: NgbModalRef = this.ngbModal.open(CopyCourseModalComponent);
+      modalRef.componentInstance.newCourseId = courseId;
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:270]: Lets separate these inits into a helper function

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
+import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
 import { SimpleModalService } from '../../../services/simple-modal.service';
 import { StatusMessageService } from '../../../services/status-message.service';
 import { StudentService } from '../../../services/student.service';
 import { TableComparatorService } from '../../../services/table-comparator.service';
 import {
   Course,
   CourseArchive,
-  Courses,
+  Courses, FeedbackQuestion, FeedbackQuestions, FeedbackSession, FeedbackSessions,
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:16]: I think we should have consistent newlines

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+})
+export class CopyCourseModalComponent implements OnInit {
+
+  // const
+  COURSE_ID_MAX_LENGTH: number = COURSE_ID_MAX_LENGTH;
+
+  @Input()
+  sessionsInCourse: FeedbackSession[] = [];
+
+  timezones: Timezone[] = [];
```
Comment (daongochieu2810) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:36]: Its probably better to group these attributes into a type e.g CopyCourseModalAttributes

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

**Comment by damithc** [comment]
@LeeBadal any plans to continue this PR? It would be a pity to throw away all the good work you've done.

**Comment by damithc** [comment]
Closing this as the work is continued in #11168

**Pull request closed without merging**